java/rocksjni/parsed_entry_info.cc

```diff
@@ -0,0 +1,191 @@
+    JNIEnv * /*env*/, jclass /*cls*/) {
+  ROCKSDB_NAMESPACE::ParsedEntryInfo *parsed_entry_info =
+      new ROCKSDB_NAMESPACE::ParsedEntryInfo();
+  return GET_CPLUSPLUS_POINTER(parsed_entry_info);
+}
+
+/*
+ * Class:     org_rocksdb_ParsedEntryInfo
+ * Method:    parseEntry
+ * Signature: (JJ[BI)V
+ */
+void JNICALL Java_org_rocksdb_ParsedEntryInfo_parseEntry(
+    JNIEnv *env, jclass /*cls*/, jlong handle, jlong options_handle,
+    jbyteArray jtarget, jint len) {
+  auto *options =
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  auto *parsed_entry_info =
+      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
+  jbyte *target = env->GetByteArrayElements(jtarget, nullptr);
+  if (target == nullptr) {
+    ROCKSDB_NAMESPACE::OutOfMemoryErrorJni::ThrowNew(env,
+           "Memory allocation failed in RocksDB JNI function");
+    return;
+  }
+  char* charCopy = new char[len];
+  std::memcpy(charCopy, target, len);
+  env->ReleaseByteArrayElements(jtarget, target, JNI_ABORT);
+  ROCKSDB_NAMESPACE::Slice target_slice(reinterpret_cast<char *>(target), len);
+  ROCKSDB_NAMESPACE::ParseEntry(target_slice, options->comparator,
+                                parsed_entry_info, true);
+}
+
+/*
+ * Class:     org_rocksdb_ParsedEntryInfo
+ * Method:    parseEntryDirect
+ * Signature: (JJLjava/nio/ByteBuffer;II)V
+ */
+void JNICALL Java_org_rocksdb_ParsedEntryInfo_parseEntryDirect(
+    JNIEnv *env, jclass /*clz*/, jlong handle, jlong options_handle,
+    jobject jbuffer, jint jbuffer_off, jint jbuffer_len) {
+  auto *options =
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  auto *parsed_entry_info =
+      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
+  auto parse = [&parsed_entry_info,
+                &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
+    ROCKSDB_NAMESPACE::ParseEntry(target_slice, options->comparator,
+                                  parsed_entry_info);
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_direct(parse, env, jbuffer, jbuffer_off,
+                                          jbuffer_len);
+}
+
+/*
+ * Class:     org_rocksdb_ParsedEntryInfo
+ * Method:    parseEntryByteArray
+ * Signature: (JJ[BII)V
+ */
+void JNICALL Java_org_rocksdb_ParsedEntryInfo_parseEntryByteArray(
+    JNIEnv *env, jclass /*clz*/, jlong handle, jlong options_handle,
+    jbyteArray jtarget, jint joff, jint jlen) {
+  auto *options =
+      reinterpret_cast<const ROCKSDB_NAMESPACE::Options *>(options_handle);
+  auto *parsed_entry_info =
+      reinterpret_cast<ROCKSDB_NAMESPACE::ParsedEntryInfo *>(handle);
+  auto parse = [&parsed_entry_info,
+                &options](ROCKSDB_NAMESPACE::Slice &target_slice) {
+    ROCKSDB_NAMESPACE::ParseEntry(target_slice, options->comparator,
+                                  parsed_entry_info);
+  };
+  ROCKSDB_NAMESPACE::JniUtil::k_op_indirect(parse, env, jtarget, joff, jlen);
```

Comment:
The slice data constructed by `k_op_indirect` has a lifetime which ends with the `k_op_indirect` call (`std::unique_ptr`), so the `parsed_entry_info` cannot be guaranteed to persist when it is returned. I wonder if it would be easier to make the `ParsedEntryInfo` a complete Java object, with members for key , type, sequence number, and fully instantiate this using JNI from a temporary C++ ParsedEntryInfo.

I also don't like the `copied_user_key` flag in `ParsedEntryInfo`, and I think if you refactored in the way I suggest, it would allow you to remove `copied_user_key`.